### PR TITLE
Optional Iterator refactor - [MOD-8844]

### PR DESCRIPTION
**Pull request by BenGoldberger**
It contains the code for the Optional Iterator refactor.

Implementation has been taken from https://github.com/RediSearch/RediSearch/pull/5619/files

### CODE REVIEW

File: tests/cpptests/micro-benchmarks/benchmark_optional_iterator.cpp [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+
+BENCHMARK_DEFINE_F(BM_OptionalIterator, ReadWithChild)(benchmark::State &state) {
+  QueryIterator *iterator_base = createOptionalWithChild();
+
+  for (auto _ : state) {
+    IteratorStatus rc = iterator_base->Read(iterator_base);
+    if (rc == ITERATOR_EOF) {
+      iterator_base->Rewind(iterator_base);
+    }
+    benchmark::DoNotOptimize(iterator_base->current);
```
Comment (JoanFM) [inline tests/cpptests/micro-benchmarks/benchmark_optional_iterator.cpp]: what does this do?

File: tests/cpptests/micro-benchmarks/benchmark_optional_iterator.cpp [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+
+class BM_OptionalIterator : public benchmark::Fixture {
+public:
+  std::vector<t_docId> childDocIds;
+  static bool initialized;
+  const t_docId maxDocId = 100000;
+  const size_t numDocs = 50000;
+  const double weight = 1.0;
+
+  void SetUp(::benchmark::State &state) {
```
Comment (JoanFM) [inline tests/cpptests/micro-benchmarks/benchmark_optional_iterator.cpp:34]: I suggest do the creation in SetUP and Release in TearDown, and to support 2 versions, you could use template parameter of the IteratorType.

File: src/iterators/optional_iterator.c [hunk 3]
```diff
@@ -0,0 +1,109 @@
+#include "optional_iterator.h"
+#include "empty_iterator.h"
+
+static void OI_Free(QueryIterator *base) {
+  OptionalIterator *nc = (OptionalIterator *)base;
```
Comment (JoanFM) [inline src/iterators/optional_iterator.c]: can we use another name? `nc`stands for ? maybe oi?

File: src/iterators/optional_iterator.c [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+  }
+
+  if (base->lastDocId != nc->child->lastDocId) {
+    nc->base.current = nc->virt;
+  } else {
+    nc->base.current = nc->child->current;
+    nc->base.current->weight = nc->weight;
+  }
+
+  nc->base.current->docId = base->lastDocId;
```
Comment (JoanFM) [inline src/iterators/optional_iterator.c]: u can use `base`directly also here

File: src/iterators/optional_iterator.c [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+  // Increase the size by one
+  base->lastDocId++;
+
+  if (base->lastDocId > nc->child->lastDocId && !base->atEOF) {
+    IteratorStatus rc = nc->child->Read(nc->child);
+    if (rc == ITERATOR_TIMEOUT) return rc;
+  }
+
+  if (base->lastDocId != nc->child->lastDocId) {
+    nc->base.current = nc->virt;
```
Comment (JoanFM) [inline src/iterators/optional_iterator.c]: I would use `base`directly

File: tests/cpptests/test_cpp_iterator_optional.cpp [hunk 6]
```diff
@@ -0,0 +1,375 @@
...
+  OptionalIterator *oi = (OptionalIterator *)iterator_base;
+
+  // Read first document (should be virtual)
+  ASSERT_EQ(iterator_base->Read(iterator_base), ITERATOR_OK);
+  ASSERT_EQ(iterator_base->current, oi->virt);
+  ASSERT_EQ(iterator_base->current->weight, weight);
+}
+
+// Test edge cases
+class OptionalIteratorEdgeCasesTest : public ::testing::Test {
```
Comment (JoanFM) [inline tests/cpptests/test_cpp_iterator_optional.cpp]: I think I am missing a Test showing how it behaves when Child Times Out

File: tests/cpptests/micro-benchmarks/benchmark_optional_iterator.cpp (hunk 1 above)
Comment (BenGoldberger) [reply tests/cpptests/micro-benchmarks/benchmark_optional_iterator.cpp]: in theory it tells the compiler to not optimize the variable, it something that the AI tool suggested.

Comment (JoanFM) [reply tests/cpptests/micro-benchmarks/benchmark_optional_iterator.cpp]: I am not sure I would use it, or we should use it everywhere. Wdyt @GuyAv46

File: src/iterators/optional_iterator.c [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+  // Only free our virtual result, not the child's result
+  if (oi->virt) {
+    IndexResult_Free(oi->virt);
+  }
+  rm_free(base);
+}
+
+static size_t OI_NumEstimated(QueryIterator *base) {
+  OptionalIterator *oi = (OptionalIterator *)base;
+  return oi->maxDocId;
```
Comment (GuyAv46) [inline src/iterators/optional_iterator.c:17]: OK for now, but we should also consult with the wildcard child if it is set

File: src/iterators/optional_iterator.c [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+  oi->virt->docId = 0;
+  if (oi->child) {
+    oi->child->Rewind(oi->child);
+  }
+}
+
+
+
+// SkipTo for OPTIONAL iterator - Non-optimized version.
+static IteratorStatus OI_SkipTo_NO(QueryIterator *base, t_docId docId) {
```
Comment (GuyAv46) [inline src/iterators/optional_iterator.c]: Let's improve the name (`*_NO` is hard to understand, we can afford to give it the full "notOptimized" suffix

File: src/iterators/optional_iterator.c [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+    base->atEOF = true;
+    return ITERATOR_EOF;
+  }
+
+  if (docId > oi->child->lastDocId) {
+    IteratorStatus rc = oi->child->SkipTo(oi->child, docId);
+    if (rc == ITERATOR_TIMEOUT) return rc;
+  }
+
+  if (docId > 0 && docId == oi->child->lastDocId) {
```
Comment (GuyAv46) [inline src/iterators/optional_iterator.c]: `docId > 0` is always true
```suggestion
  if (docId == oi->child->lastDocId) {
```

File: src/iterators/optional_iterator.c [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+  OptionalIterator *oi = (OptionalIterator *)base;
+  if (base->atEOF || base->lastDocId >= oi->maxDocId) {
+    base->atEOF = true;
+    return ITERATOR_EOF;
+  }
+
+  // Increase the size by one
+  base->lastDocId++;
+
+  if (base->lastDocId > oi->child->lastDocId && !base->atEOF) {
```
Comment (GuyAv46) [inline src/iterators/optional_iterator.c]: `!base->atEOF` is always true here
```suggestion
  if (base->lastDocId > oi->child->lastDocId) {
```

File: src/iterators/optional_iterator.c [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+  }
+
+  base->current->docId = base->lastDocId;
+  return ITERATOR_OK;
+}
+
+// Create a new OPTIONAL iterator - Non-Optimized version.
+QueryIterator *IT_V2(NewOptionalIterator_NonOptimized)(QueryIterator *it, t_docId maxDocId, size_t numDocs, double weight) {
+  OptionalIterator *oi = rm_calloc(1, sizeof(*oi));
+  oi->child = it ? it : IT_V2(NewEmptyIterator)();
```
Comment (GuyAv46) [inline src/iterators/optional_iterator.c]:
```suggestion
  oi->child = it ?: IT_V2(NewEmptyIterator)();
```
I wonder if this case is even possible, and whether we should have an empty iterator as a child or simply return a wildcard iterator instead of the optional

File: src/iterators/optional_iterator.c [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+    base->current = oi->child->current;
+    base->current->weight = oi->weight;
+  }
+
+  base->current->docId = base->lastDocId;
+  return ITERATOR_OK;
+}
+
+// Create a new OPTIONAL iterator - Non-Optimized version.
+QueryIterator *IT_V2(NewOptionalIterator_NonOptimized)(QueryIterator *it, t_docId maxDocId, size_t numDocs, double weight) {
```
Comment (GuyAv46) [inline src/iterators/optional_iterator.c]: Let's have a "generic" constructor, the optimization should be chosen automatically when possible, and not be part of the API

File: tests/cpptests/micro-benchmarks/benchmark_optional_iterator.cpp [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+#include "redismock/util.h"
+#include "iterator_util.h"
+
+#include <random>
+#include <vector>
+
+#include "src/iterators/iterator_api.h"
+#include "src/iterators/optional_iterator.h"
+#include "src/iterators/idlist_iterator.h"
+#include "src/iterators/empty_iterator.h"
```
Comment (GuyAv46) [inline tests/cpptests/micro-benchmarks/benchmark_optional_iterator.cpp]: Why do we need them here?

File: tests/cpptests/test_cpp_iterator_optional.cpp [hunk 14]
```diff
@@ -0,0 +1,481 @@
...
+  ASSERT_EQ(iterator_base->SkipTo(iterator_base, maxDocId + 1), ITERATOR_EOF);
+  ASSERT_TRUE(iterator_base->atEOF);
+}
+
+TEST_F(OptionalIteratorEmptyTest, SkipToZero) {
+  // Test skipping to docId 0 (should set to 0)
+  ASSERT_EQ(iterator_base->SkipTo(iterator_base, 0), ITERATOR_OK);
+  ASSERT_EQ(iterator_base->current->docId, 0);
+  ASSERT_EQ(iterator_base->lastDocId, 0);
+}
```
Comment (GuyAv46) [inline tests/cpptests/test_cpp_iterator_optional.cpp]: Not required. Lets drop this "feature"

File: tests/cpptests/micro-benchmarks/benchmark_optional_iterator.cpp [hunk 15]
```diff
@@ -0,0 +1,236 @@
...
+BENCHMARK_REGISTER_F(BM_OptionalIteratorEmpty, ReadEmpty)->CHILD_DOCS_SCENARIOS();
+BENCHMARK_REGISTER_F(BM_OptionalIteratorWithChild, ReadWithChild)->CHILD_DOCS_SCENARIOS();
+BENCHMARK_REGISTER_F(BM_OptionalIteratorEmpty, SkipToEmpty)->CHILD_DOCS_SCENARIOS();
+BENCHMARK_REGISTER_F(BM_OptionalIteratorWithChild, SkipToWithChild)->CHILD_DOCS_SCENARIOS();
+BENCHMARK_REGISTER_F(BM_OptionalIteratorWithChild, SkipToRealHits)->CHILD_DOCS_SCENARIOS();
+BENCHMARK_REGISTER_F(BM_OptionalIteratorWithChild, SkipToVirtualHits)->CHILD_DOCS_SCENARIOS();
+BENCHMARK_REGISTER_F(BM_OptionalIteratorWithChild, MixedOperations)->CHILD_DOCS_SCENARIOS();
+BENCHMARK_REGISTER_F(BM_OptionalIteratorWithChild, FullIteration)->CHILD_DOCS_SCENARIOS();
+BENCHMARK_REGISTER_F(BM_OptionalIteratorWithChild, RewindPerformance)->CHILD_DOCS_SCENARIOS();
+BENCHMARK_REGISTER_F(BM_OptionalIteratorEmpty, CreateAndDestroy)->CHILD_DOCS_SCENARIOS();
```
Comment (GuyAv46) [inline tests/cpptests/micro-benchmarks/benchmark_optional_iterator.cpp]: Let's remove the `CHILD_DOCS_SCENARIOS` in the cases of `BM_OptionalIteratorEmpty`

File: tests/cpptests/micro-benchmarks/benchmark_optional_iterator.cpp [hunk 16]
```diff
@@ -0,0 +1,236 @@
...
+    // Just advance the iterator
+  }
+
+  for (auto _ : state) {
+    iterator_base->Rewind(iterator_base);
+    benchmark::DoNotOptimize(iterator_base->lastDocId);
+  }
+}
+
+BENCHMARK_DEFINE_F(BM_OptionalIteratorEmpty, CreateAndDestroy)(benchmark::State &state) {
```
Comment (GuyAv46) [inline tests/cpptests/micro-benchmarks/benchmark_optional_iterator.cpp]: We shouldn't care for this scenario

File: tests/cpptests/micro-benchmarks/benchmark_optional_iterator.cpp [hunk 17]
```diff
@@ -0,0 +1,236 @@
...
+    // Iterate through all documents
+    while (temp_iterator->Read(temp_iterator) == ITERATOR_OK) {
+      benchmark::DoNotOptimize(temp_iterator->current->docId);
+    }
+
+    temp_iterator->Free(temp_iterator);
+  }
+}
+
+BENCHMARK_DEFINE_F(BM_OptionalIteratorWithChild, RewindPerformance)(benchmark::State &state) {
```
Comment (GuyAv46) [inline tests/cpptests/micro-benchmarks/benchmark_optional_iterator.cpp]: We shouldn't care for this scenario

File: tests/cpptests/micro-benchmarks/benchmark_optional_iterator.cpp [hunk 18]
```diff
@@ -0,0 +1,236 @@
...
+
+    if (rc == ITERATOR_EOF) {
+      iterator_base->Rewind(iterator_base);
+    }
+
+    benchmark::DoNotOptimize(rc);
+  }
+}
+
+BENCHMARK_DEFINE_F(BM_OptionalIteratorWithChild, FullIteration)(benchmark::State &state) {
```
Comment (GuyAv46) [inline tests/cpptests/micro-benchmarks/benchmark_optional_iterator.cpp]: Does this scenario give interesting results? I don't think we should care about it

File: tests/cpptests/micro-benchmarks/benchmark_optional_iterator.cpp [hunk 19]
```diff
@@ -0,0 +1,236 @@
...
+      virtualIndex = 0;
+    }
+
+    t_docId target = virtualDocIds[virtualIndex++];
+    IteratorStatus rc = iterator_base->SkipTo(iterator_base, target);
+    benchmark::DoNotOptimize(rc);
+  }
+}
+
+BENCHMARK_DEFINE_F(BM_OptionalIteratorWithChild, MixedOperations)(benchmark::State &state) {
```
Comment (GuyAv46) [inline tests/cpptests/micro-benchmarks/benchmark_optional_iterator.cpp]: Do we need this scenario? Isn't it just the average of `Read` and `SkipTo`?

File: tests/cpptests/micro-benchmarks/benchmark_optional_iterator.cpp [hunk 20]
```diff
@@ -0,0 +1,236 @@
...
+  for (auto _ : state) {
+    IteratorStatus rc = iterator_base->SkipTo(iterator_base, docId);
+    docId += 100;
+    if (rc == ITERATOR_EOF || docId > maxDocId) {
+      iterator_base->Rewind(iterator_base);
+      docId = 1;
+    }
+    benchmark::DoNotOptimize(rc);
+  }
+}
```
Comment (GuyAv46) [inline tests/cpptests/micro-benchmarks/benchmark_optional_iterator.cpp]:
```suggestion
BENCHMARK_DEFINE_F(BM_OptionalIteratorWithChild, SkipToWithChild)(benchmark::State &state) {
  t_offset step = 10;
  for (auto _ : state) {
    IteratorStatus rc = iterator_base->SkipTo(iterator_base, iterator_base->lastDocId + step);
    if (rc == ITERATOR_EOF) {
      iterator_base->Rewind(iterator_base);
    }
    benchmark::DoNotOptimize(rc);
  }
}
```

File: tests/cpptests/micro-benchmarks/benchmark_optional_iterator.cpp (hunk 2 above)
Comment (GuyAv46) [reply tests/cpptests/micro-benchmarks/benchmark_optional_iterator.cpp:34]: Either with templates or set both new and old in `SetUp`. Eventually we will delete the old one

File: src/iterators/optional_iterator.c (hunk 9 above)
Comment (BenGoldberger) [reply src/iterators/optional_iterator.c]: in theory yes, but in we dont enforce it in the test, so it leads to segfault if not checked

File: src/iterators/optional_iterator.c (hunk 7 above)
Comment (BenGoldberger) [reply src/iterators/optional_iterator.c:17]: since the non optimize version doesnt use the wildcard iterator I left it for now

File: src/iterators/optional_iterator.c (hunk 9 above)
Comment (GuyAv46) [reply src/iterators/optional_iterator.c]: So let's add an assertion (that `docId` > `base->lastDocId`) and drop this validation

Comment (JoanFM) [reply src/iterators/optional_iterator.c]: agree here with Guy

File: src/iterators/optional_iterator.c [hunk 21]
```diff
@@ -0,0 +1,109 @@
...
+  oi->virt->docId = 0;
+  if (oi->child) {
+    oi->child->Rewind(oi->child);
+  }
+}
+
+
+
+// SkipTo for OPTIONAL iterator - Non-optimized version.
+static IteratorStatus OI_SkipTo_NotOptimized(QueryIterator *base, t_docId docId) {
```
Comment (JoanFM) [inline src/iterators/optional_iterator.c:35]: can you add a comment up here, explaining the logic? In this SkipTo, there is no ITERATOR_NOTFOUND? It always matches?

File: src/iterators/optional_iterator.c [hunk 22]
```diff
@@ -0,0 +1,109 @@
...
+
+// Read from an OPTIONAL iterator - Non-Optimized version.
+static IteratorStatus OI_ReadSorted_NotOptimized(QueryIterator *base) {
+  OptionalIterator *oi = (OptionalIterator *)base;
+  if (base->atEOF || base->lastDocId >= oi->maxDocId) {
+    base->atEOF = true;
+    return ITERATOR_EOF;
+  }
+
+  // Increase the size by one
```
Comment (JoanFM) [inline src/iterators/optional_iterator.c]: it is not the size

Comment (JoanFM) [reply src/iterators/optional_iterator.c]: maybe, point to the next one?

File: src/iterators/optional_iterator.c [hunk 23]
```diff
@@ -0,0 +1,109 @@
...
+  base->lastDocId++;
+
+  if (base->lastDocId > oi->child->lastDocId) {
+    IteratorStatus rc = oi->child->Read(oi->child);
```
Comment (JoanFM) [inline src/iterators/optional_iterator.c:76]: Is reading one time enough? Or should we read until lastDocID is surpassed?

Comment (JoanFM) [reply src/iterators/optional_iterator.c:76]: If we Read until lastDocId is surpassed I am not sure how we would behave in next iteration

File: tests/cpptests/index_utils.h [hunk 24]
```diff
@@ -47,3 +48,32 @@ size_t CalculateNumericInvertedIndexMemory(NumericRangeTree *rt, NumericRangeNod
 size_t NumericRangeGetMemory(const NumericRangeNode *Node);
 
 NumericRangeTree *getNumericTree(IndexSpec *spec, const char *field);
+
+class MockQueryEvalCtx {
```
Comment (JoanFM) [inline tests/cpptests/index_utils.h:52]: I think for the tests is better and more explanatory to have a constructor where you can inject the Iterator u want than having all these mocks that are quite hard to follow. Then you can do some testing on the constructor if needed.

File: src/iterators/optional_iterator.c (hunk 23 above)
Comment (BenGoldberger) [reply src/iterators/optional_iterator.c:76]: we would return `oi->virt` until child->lastDocId

File: tests/cpptests/index_utils.h (hunk 24 above)
Comment (BenGoldberger) [reply tests/cpptests/index_utils.h:52]: I'm not sure I follow. do you want me to add a constructor for the mock?

Comment (JoanFM) [reply tests/cpptests/index_utils.h:52]: check here https://github.com/RediSearch/RediSearch/pull/6307

I wonder if it is easier to follow to have just some constructor that u can use to inject your iterator, instead of trying to mock all these structures.

Then you can consider what u may see best.

File: src/iterators/optional_iterator.c (hunk 21 above)
Comment (BenGoldberger) [reply src/iterators/optional_iterator.c:35]: Yes similar to the wildcard iterator

File: tests/cpptests/index_utils.h (hunk 24 above)
Comment (BenGoldberger) [reply tests/cpptests/index_utils.h:52]: I think this implementation  is quite simple for this benchmark

File: tests/cpptests/test_cpp_iterator_optional.cpp [hunk 25]
```diff
@@ -0,0 +1,404 @@
...
+  // Should be virtual hit
+  OptionalIterator *oi = (OptionalIterator *)iterator_base;
+  ASSERT_EQ(iterator_base->current, oi->virt);
+}
+
+TEST_F(OptionalIteratorTest, SkipToSequence) {
+  // Test skipping to various docIds in sequence
+  t_docId targets[] = {5, 15, 25, 35, 45, 55, 65, 75, 85, 95};
+
+  for (size_t i = 0; i < sizeof(targets) / sizeof(targets[0]); i++) {
```
Comment (GuyAv46) [inline tests/cpptests/test_cpp_iterator_optional.cpp]:
```suggestion
  for (auto &target : targets) {
```

**Review by GuyAv46 — DISMISSED** [review]

File: src/iterators/optional_iterator.c (hunk 3 above)
Comment (GuyAv46) [reply src/iterators/optional_iterator.c]: `nc` stands for "not context". Sowing the history of where it was copied from :)

File: src/iterators/optional_iterator.c [hunk 26]
```diff
@@ -0,0 +1,112 @@
...
+  }
+}
+
+
+
+// SkipTo for OPTIONAL iterator - Non-optimized version.
+// Skip to a specific docId. If the child has a hit on this docId, return it.
+// Otherwise, return a virtual hit.
+static IteratorStatus OI_SkipTo_NotOptimized(QueryIterator *base, t_docId docId) {
+  assert(docId > 0);
```
Comment (JoanFM) [inline src/iterators/optional_iterator.c]: RS_ASSERT I believe is best. I guess somehow we skip the assertion in prod

File: src/iterators/optional_iterator.c [hunk 27]
```diff
@@ -0,0 +1,112 @@
...
+
+  // Point to next doc
+  base->lastDocId++;
+
+  if (base->lastDocId > oi->child->lastDocId) {
+    IteratorStatus rc = oi->child->Read(oi->child);
+    if (rc == ITERATOR_TIMEOUT) return rc;
+  }
+
+  if (base->lastDocId != oi->child->lastDocId) {
```
Comment (JoanFM) [inline src/iterators/optional_iterator.c]: very nitpicky, ignore if u want, but what about having positive logical here as in the SkipTo?

Also, why is the oi-virt-docId not updated?

File: tests/cpptests/test_cpp_iterator_optional.cpp [hunk 28]
```diff
@@ -0,0 +1,404 @@
...
+  // After reading all docs, should return EOF
+  ASSERT_EQ(iterator_base->Read(iterator_base), ITERATOR_EOF);
+  ASSERT_TRUE(iterator_base->atEOF);
+}
+
+TEST_F(OptionalIteratorWithEmptyChildTest, SkipToVirtualHits) {
+  // Skip to various docIds - all should be virtual hits
+  t_docId targets[] = {5, 15, 25, 35, 45};
+
+  for (size_t i = 0; i < sizeof(targets) / sizeof(targets[0]); i++) {
```
Comment (JoanFM) [inline tests/cpptests/test_cpp_iterator_optional.cpp]: same as Guy's comment on target loops.

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: src/iterators/optional_iterator.c (hunk 26 above)
Comment (GuyAv46) [reply src/iterators/optional_iterator.c]: Good catch! I'd also change it to validate that the requested ID is ahead of the last ID
```suggestion
  RS_ASSERT(docId > base->lastDocId);
```

File: src/iterators/optional_iterator.c (hunk 27 above)
Comment (BenGoldberger) [reply src/iterators/optional_iterator.c]: it probably should be updated

**Review by JoanFM — APPROVED** [review]

**Review by GuyAv46 — APPROVED** [review]

File: src/iterators/optional_iterator.c [hunk 29]
```diff
@@ -0,0 +1,112 @@
...
+  } else {
+    oi->virt->docId = base->lastDocId;
+    base->current = oi->virt;
+  }
+  return ITERATOR_OK;
+}
+
+// Create a new OPTIONAL iterator - Non-Optimized version.
+QueryIterator *IT_V2(NewOptionalIterator)(QueryIterator *it, t_docId maxDocId, size_t numDocs, double weight) {
+  assert(it != NULL);
```
Comment (GuyAv46) [inline src/iterators/optional_iterator.c:92]: On a second look, this may happen (for example, when searching for `~foo` but no document has `foo` so there is no inverted index for it)
I suggest we return a wildcard iterator in this case.
We have to check the implications on `FT.PROFILE` and `FT.EXPLAIN` when we start using the new iterators

Comment (GuyAv46) [reply src/iterators/optional_iterator.c:92]: alternativly, use `RS_ASSERT` and we will have to handle this case when using the API
